### PR TITLE
Calculate position ids in modeling utils for all generative models

### LINKED ISSUE #29149: Generate: support passing position_ids
Thank you @tengomucho, for uncovering this bug.

### The problem

In a nutshell, passing the correct `position_ids` to `generate` should result in exactly the same results as not passing them. In other words, the following test should pass on all models, if added to `GenerationTesterMixin`. We can see that it is failing in general.

```py
    def test_passing_position_ids(self):
        # Check that passing position ids to generate yields the same results as not passing them, if the position ids
        # are correctly built. If the test fails, it means one of two things:
        # 1 - the manual position ids are not being piped correctly; OR
        # 2 - the automated position ids are not being correctly built.
        for model_class in self.all_generative_model_classes:
            config, input_ids, attention_mask, _ = self._get_input_ids_and_config(batch_size=1)
            if config.is_encoder_decoder:
                self.skipTest("This model does not support position_ids")

            # To truly test this property, let's create a batch where the second row corresponds to the test input with
            # left padding of 1.
            pad_token = torch.tensor([[config.pad_token_id or 0]], device=input_ids.device, dtype=input_ids.dtype)
            input_ids = torch.cat((input_ids, torch.cat((pad_token, input_ids[:, 1:]), dim=1)), dim=0)
            pad_mask = torch.zeros((1, 1), dtype=attention_mask.dtype, device=attention_mask.device)
            attention_mask = torch.cat((attention_mask, torch.cat((pad_mask, attention_mask[:, 1:]), dim=1)), dim=0)
            position_ids = torch.clamp(torch.cumsum(attention_mask, dim=-1) - 1, min=0)

            config.use_cache = True
            config.is_decoder = True

            model = model_class(config).to(torch_device).eval()

            try:
                output_position_ids = model.generate(
                    input_ids,
                    attention_mask=attention_mask,
                    position_ids=position_ids,
                    max_new_tokens=10
                )
            except ValueError as exc:
                if "The following `model_kwargs` are not used by the model: ['position_ids']" in str(exc):
                    self.skipTest("This model does not support position_ids")
                else:
                    raise
            output_no_position_ids = model.generate(
                input_ids,
                attention_mask=attention_mask,
                max_new_tokens=10
            )
            self.assertListEqual(output_no_position_ids.tolist(), output_position_ids.tolist())
```

### The fix

There are two root causes for this:
1. `position_ids` is rejected in some models when it is passed (e.g. see [here](https://github.com/huggingface/transformers/blob/3c00b885b92fbcd0e7451e56ccf424a2d5a19bbb/src/transformers/models/gpt2/modeling_gpt2.py#L1022)). These models often assume no padding when `position_ids` is rejected.
2. `position_ids` is never updated, so it is only correct when created from scratch (=not passed).

As such, a fix to this problem should consist in updating `position_ids` in `generate`, with `prepare_inputs_for_generation` only creating new `position_ids` when they don't exist.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As it was discussed under [this PR](https://github.com/huggingface/transformers/pull/29583), position ids in some models are not calculated/inferred from attn mask in `forward`, which gives incorrect positions when the inputs is left padded.

To be consistent and for ease of maintaining, the logic of inferring position ids is moved to "modeling_utils.py" and all generative models call that method in their `forward` and `prepare_inputs_for_generation`. I added two tests, to check whether model outputs are same when position ids are passed by a user vs. when inferred from input ids or embeds. 

Also Fixes #29149.

The newly added tests are passing. Plus slow tests on vision models, because they still do not have GenerationTesterMixin.

Btw, I see that non-generative models already use `create_position_ids_from_input_ids` method which is copied separately in each model's file. The logic is a bit different from generative models, because they start counting from "padding_idx" and not "0". Anyway, I guess it is still possible to merge that method and the one proposed here, to have one "get_position_id" for all models in the "modeling_utils". 
@gante WDYT ?

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
About the framework changes: I found that tf/flax has a slightly different way to get `position_ids` from torch models. Those frameworks generate position ids in `forward` without taking into account attention mask, same thing we had in torch before these fixes. 

I made tf and flax same way as torch is now with a cumsum over attention mask, so that the equivalence over frameworks tests pass. I am not sure if we need similar test for tf/flax to "test_position_ids". Tests should pass now, at least locally it seemed okay

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1607,6 +1667,56 @@ def test_generate_continue_from_past_key_values(self):
...
+            model_forward_args = inspect.signature(model.forward).parameters
+            if "position_ids" not in model_forward_args:
+                self.skipTest("This model doesn't use `position_ids`")
+
+            out_wo_positions = model.generate(input_ids, attention_mask=attention_mask, max_new_tokens=5)
+
+            # infer position ids from attn mask and generate again
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            position_ids = position_ids[..., -input_ids.shape[-1] :].view(-1, input_ids.shape[-1])
```
Comment (gante) [inline tests/generation/test_utils.py:1683]: Use the newly created function instead? The main goal of this test is to ensure we can pass `position_ids`, reusing the function can save us future trouble here :)

(a small unit test with numerical checks outside the test mixins for the newly created function would also be nice!)

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -1607,6 +1667,56 @@ def test_generate_continue_from_past_key_values(self):
...
+
+            inputs_embeds = model.get_input_embeddings()(input_ids)
+            out_wo_positions = model.generate(
+                inputs_embeds=inputs_embeds, attention_mask=attention_mask, max_new_tokens=5
+            )
+
+            # infer position ids from attn mask and generate again
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            position_ids = position_ids[..., -input_ids.shape[-1] :].view(-1, input_ids.shape[-1])
```
Comment (gante) [inline tests/generation/test_utils.py:1711]: (same comment as above)

**Review by gante — COMMENTED** [review]
In general looks good, thank you for tackling this refactor 💪 

A few notes:
1. No TF function to infer the position IDs? 😢 TF feels neglected 💔  
2. There are CI errors in the model equivalence. Model equivalence is flaky by nature, make sure you run model equivalence for all models with flake finder locally!
3. After you're happy with the changes, commit with `[test_all]` and tag me again. I've glanced over the model-level changes after the first few models, I'll do a final check more carefully after the full CI is green 🤗

File: src/transformers/models/codegen/modeling_codegen.py [hunk 3]
```diff
@@ -467,8 +467,10 @@ def forward(
             past_length = past_key_values[0][0].size(-2)
 
         if position_ids is None:
-            position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
-            position_ids = position_ids.unsqueeze(0)
+            seq_length = input_ids.shape[-1] if input_ids is not None else inputs_embeds.shape[1]
+            position_ids = self.get_position_ids_from_attention_mask(
+                attention_mask, past_length, seq_length=seq_length, device=device
+            )
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:473]: yes, good point

**Comment by zucchini-nlp** [comment]
@gante the comments are addressed now. TF cannot have the "get_position_ids" method in PretrainedModel because all input related preparations in TF happen in a "keras.layers.Layer" class. I am not sure if we can or should be moving the position_id preparation into the "PretrainedModel", since there are only 3 TF models that were needed change.

Also, to note for Flax-based encoder-decoder models: the attention mask for decoder part is overriden to be full, because when using decoder-only model as decoder part the position ids are calculated differently (I mean only the unattended part). In random initialized models it is causing logits mismatch, even if the attention masks out unattended positions. In pre-trained models that does not happen 🤔

File: src/transformers/generation/candidate_generator.py [hunk 4]
```diff
@@ -423,3 +424,18 @@ def _prepare_token_type_ids(model_kwargs: Dict[str, Any], new_length: int) -> Di
...
+    position_ids = model_kwargs.get("position_ids")
+    if position_ids is None:
+        return model_kwargs
+
+    # we assume batch_size=1 for assited decoding (needs rework if bs > 1)
+    length_diff = new_length - position_ids[0, -1]
+    if length_diff < 0:
+        position_ids = position_ids[:, :length_diff]
+    elif length_diff > 0:
+        new_position_ids = torch.arange(position_ids[0, -1], new_length, device=position_ids.device).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/candidate_generator.py:439]: ^ this function still needs better variable names and/or a docstring

File: src/transformers/models/codegen/modeling_codegen.py [hunk 5]
```diff
@@ -614,12 +614,16 @@ def prepare_inputs_for_generation(self, input_ids, inputs_embeds=None, past_key_
...
+        seq_length = (
+            inputs_embeds.shape[1] if inputs_embeds is not None and past_key_values is None else input_ids.shape[1]
+        )
+        if position_ids is None:
+            device = input_ids.device if input_ids is not None else inputs_embeds.device
+            position_ids = self.get_position_ids_from_attention_mask(
+                attention_mask, past_length, seq_length=seq_length, device=device
+            )
+        else:
+            position_ids = position_ids[:, -seq_length:]
```
Comment (gante) [inline src/transformers/models/codegen/modeling_codegen.py:626]: I think we can remove all this code, actually 👀 I see the following cases:
1. `position_ids is None` -> the forward pass correctly computes `position_ids`, due to the changes in this PR
2. `position_ids is not None` -> the user has defined `position_ids`, it's its own responsibility to pass them correctly

WDYT? (this logic would apply to all models, and would make maintenance easier for us 👼 )

File: src/transformers/models/encoder_decoder/modeling_encoder_decoder.py [hunk 6]
```diff
@@ -593,6 +593,7 @@ def forward(
             argument[len("decoder_") :]: value for argument, value in kwargs.items() if argument.startswith("decoder_")
         }
 
+        print(self.encoder)
```
Comment (gante) [inline src/transformers/models/encoder_decoder/modeling_encoder_decoder.py:596]:
```suggestion
```

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 7]
```diff
@@ -702,7 +709,9 @@ def prepare_inputs_for_generation(self, inputs, past_key_values=None, use_cache=
         attention_mask = kwargs.get("attention_mask", None)
 
         if attention_mask is not None and position_ids is None:
-            position_ids = tf.math.cumsum(attention_mask, axis=-1, exclusive=True)
```
Comment (gante) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:705]: this one should be correct, no? 🤔 

(the same comment applies to other TF models)

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 8]
```diff
@@ -344,8 +344,15 @@ def call(
...
-            position_ids = tf.tile(position_ids, [input_shape[0], 1])
+            if attention_mask is not None:
+                position_ids = tf.cumsum(tf.cast(attention_mask, tf.int64), axis=-1) - 1
+                # create ones tensor to match dtypes, otherwise we get errors
+                ones_tensor = tf.ones_like(position_ids, dtype=tf.int64)
+                position_ids = tf.where(attention_mask == 0, ones_tensor, position_ids)
+                position_ids = position_ids[..., -input_shape[-1] :]
+                position_ids = tf.reshape(position_ids, (-1, input_shape[-1]))
+            else:
+                position_ids = tf.expand_dims(tf.range(past_length, input_shape[-1] + past_length), axis=0)
```
Comment (gante) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:355]:
```suggestion
                position_ids = tf.math.cumsum(attention_mask, axis=-1, exclusive=True)
```
(see comment below)

Comment (gante) [reply src/transformers/models/ctrl/modeling_tf_ctrl.py:355]: the same logic applies to other TF models

File: tests/models/encoder_decoder/test_modeling_flax_encoder_decoder.py [hunk 9]
```diff
@@ -299,6 +301,14 @@ def check_pt_flax_equivalence(self, pt_model, fx_model, inputs_dict):
         flax_inputs = inputs_dict
         pt_inputs = {k: torch.tensor(v.tolist()) for k, v in flax_inputs.items()}
 
+        # when model weights are random init masking with attn_mask still leads to logits
+        # mismatch, which does not happen if pre-trained models are used. That causes error in encoder-decoder models
+        # when decoder_only is used in as backbone (GPT2), because GPT prepares positions depending on attn mask (for torch)
+        # and as arange in flax. That's why we init attn mask with all `1`
+        if "decoder_attention_mask" in pt_inputs:
+            pt_inputs["decoder_attention_mask"] = torch.ones_like(pt_inputs["decoder_attention_mask"])
+            inputs_dict["decoder_attention_mask"] = jnp.ones_like(inputs_dict["decoder_attention_mask"])
```
Comment (gante) [inline tests/models/encoder_decoder/test_modeling_flax_encoder_decoder.py:310]: This change should no longer be needed, correct?

(as a general rule, we shouldn't fudge these equivalence tests :) )

File: tests/models/gemma/test_modeling_flax_gemma.py [hunk 10]
```diff
@@ -143,6 +143,9 @@ def check_use_cache_forward_with_attn_mask(self, model_class_name, config, input
         max_decoder_length = 20
         model = model_class_name(config)
 
+        # make full attn mask since below we are preparing position ids assuming it's all ones
+        attention_mask = jnp.ones_like(attention_mask)
```
Comment (gante) [inline tests/models/gemma/test_modeling_flax_gemma.py:147]: the other way around: we should update the creation of `position_ids` (below) to match the mask

The same comment applies to other FLAX test changes

File: tests/models/llama/test_modeling_flax_llama.py [hunk 11]
```diff
@@ -149,8 +149,8 @@ def check_use_cache_forward_with_attn_mask(self, model_class_name, config, input
         )
 
         past_key_values = model.init_cache(input_ids.shape[0], max_decoder_length)
-        position_ids = jnp.broadcast_to(
-            jnp.arange(input_ids.shape[-1] - 1)[None, :], (input_ids.shape[0], input_ids.shape[-1] - 1)
+        position_ids = model.get_position_ids_from_attention_mask(
```
Comment (gante) [inline tests/models/llama/test_modeling_flax_llama.py:152]: yes, like this!

File: tests/models/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py [hunk 12]
```diff
@@ -414,6 +414,14 @@ def check_pt_flax_equivalence(self, pt_model, fx_model, inputs_dict):
...
         pt_inputs = {k: torch.tensor(v.tolist()) for k, v in flax_inputs.items()}
 
+        # when model weights are random init masking with attn_mask still leads to logits
+        # mismatch, which does not happen if pre-trained models are used. That causes error in encoder-decoder models
+        # when decoder_only is used in as backbone (GPT2), because GPT prepares positions depending on attn mask (for torch)
+        # and as arange in flax. That's why we init attn mask with all `1`
+        if "decoder_attention_mask" in pt_inputs:
+            pt_inputs["decoder_attention_mask"] = torch.ones_like(pt_inputs["decoder_attention_mask"])
+            inputs_dict["decoder_attention_mask"] = jnp.ones_like(inputs_dict["decoder_attention_mask"])
+
```
Comment (gante) [inline tests/models/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py:424]: same comment as in the same pattern above, we should remove this

**Review by gante — COMMENTED** [review]
A few more pattern fixes and should be ready to go 🤞

**Comment by zucchini-nlp** [comment]
Will reopen this one later, as a new PR. It will need resolving merge conflicts and propagating changes to new models + PR comments.
